src/authorization/WebAclAuthorizer.ts

```diff
@@ -81,28 +87,88 @@ export class WebAclAuthorizer extends Authorizer {
+    let agentPermissions: PermissionSet;
+    if (this.isAuthenticated(agent) && this.accessCheckers !== undefined) {
+      agentPermissions = await this.determineAgentPermissions(acl, agent, publicPermissions);
 
-    return new WebAclAuthorization(userPermissions, publicPermissions);
+      if (agentPermissions.write) {
+        // Write permission implies Append permission
+        agentPermissions.append = true;
+      }
+    } else {
+      agentPermissions = { ...publicPermissions } as PermissionSet;
+    }
+
+    return new WebAclAuthorization(agentPermissions, publicPermissions);
   }
 
-  /**
-   * Determines the available permissions for the given credentials.
-   * @param credentials - Credentials to find the permissions for.
-   * @param acl - Store containing all relevant authorization triples.
-   */
-  private determinePermissions(credentials: Credentials, acl: Store): PermissionSet {
+  private modesMapper(mode: string): (keyof PermissionSet) | undefined {
+    switch (mode) {
+      case ACL.Read: return 'read';
+      case ACL.Write: return 'write';
+      case ACL.Append: return 'append';
+      case ACL.Control: return 'control';
+      default: return undefined;
+    }
+  }
+
+  private determinePublicPermissions(acl: Store): PermissionSet {
     const permissions: PermissionSet = {
       read: false,
       write: false,
       append: false,
       control: false,
     };
-    for (const mode of (Object.keys(permissions) as (keyof PermissionSet)[])) {
-      permissions[mode] = this.hasPermission(credentials, acl, mode);
+
+    const publicAcl = new Store();
+    // Keep only unique values
+    const publicRules = new Set(acl.getSubjects(ACL.agentClass, FOAF.Agent, null));
+
+    publicRules.forEach((rule): any => publicAcl.addQuads(acl.getQuads(rule, null, null, null)));
+    // Keep only unique values
+    const publicModes = new Set(publicAcl.getObjects(null, ACL.mode, null));
+
+    publicModes.forEach((mode): void => {
+      const property = this.modesMapper(mode.value);
+      if (typeof property !== 'undefined') {
+        permissions[property] = true;
+      }
+    });
+
+    return permissions;
+  }
+
+  private async determineAgentPermissions(acl: Store, agent: Credentials,
+    publicPermissions: PermissionSet): Promise<PermissionSet> {
+    const permissions = { ...publicPermissions } as PermissionSet;
+
+    // Keep only unique values
+    const rules = new Set(acl.getSubjects(null, null, null));
+
+    for (const rule of rules) {
+      const hasAccess = await promiseAny(this.accessCheckers!
+        .map(async(handler): Promise<boolean> => await handler.handleSafe({ acl, rule, agent })));
+
+      if (hasAccess) {
+        const modes = acl.getObjects(rule, ACL.mode, null);
+        modes.forEach((mode): void => {
+          const property = this.modesMapper(mode.value);
+          if (typeof property !== 'undefined') {
+            permissions[property] = true;
+          }
+        });
+      }
     }
+
```

Comment:
I'm a bit surprised by the major rewrite here. I assume this is for performance; do we have hard evidence that it is faster?